---
title: "Session-02"
output:
  revealjs::revealjs_presentation:
    reveal_options:
      width: "100%"
      height: "120%"
      margin: 0.1
      minScale: 0.2
      maxScale: 1
---

```{=html}
<!-- <style> -->
<!--     .reveal section p { -->
<!--     display: inline-block; -->
<!--     font-size: 0.6em; -->
<!--     line-height: 1.2em; -->
<!--     vertical-align: top; -->
<!--   } -->
<!--     .reveal section li { -->
<!--     display: inline-block; -->
<!--     font-size: 0.6em; -->
<!--     line-height: 1.2em; -->
<!--     vertical-align: top; -->
<!--   } -->
<!-- </style> -->
```
# Starting an analysis project

## Know the project goals

-   Experimental design
    -   Library type, sequencing depth, targeted \# of cells
-   Relevant sample metadata
    -   Conditions, timepoints
-   Celltypes of interest
    -   Niche populations, heterogenous vs sorted samples

## Organizing bioinformatic analysis

Don't

-   Use a single analysis script
-   Make undocumented changes to data via the console or other interative tools
-   Become attached to data in your environment

Do

-   Split your workflow into stages
-   Save intermediate objects between processing steps
-   Periodically rerun your scripts

If you're ever afraid of not being able to reproduce an outcome, reassess your workflows

## Tools for reproducible workflows

-   [r projects](https://r4ds.hadley.nz/workflow-scripts.html#projects)
    -   Isolate analyses, organize filepaths
-   [renv](https://rstudio.github.io/renv/index.html)
    -   Record and manage package versions
-   [Git](https://happygitwithr.com/)
    -   Track and revert changes; backup and share code

# Cellranger output files

Data may be stored in matrix/tsv formats or [HDF5](https://www.loc.gov/preservation/digital/formats/fdd/fdd000229.shtml) format

Matrix format

<small>`barcodes.tsv.gz  features.tsv.gz  matrix.mtx.gz`</small>

H5 format

<small>`sample_feature_bc_matrix.h5`</small>

Running `cellranger multi` with other assays (e.g. T or B cell repertoire seq) may produce other output files

# Single-cell analysis packages

[Seurat](https://satijalab.org/seurat/) - R based, probably most popular, lots tools built around Seurat's framework

[Scanpy](https://scanpy.readthedocs.io/en/stable/) - Python based, newer, more amenable to trajectory analyses

[SeqGeq](https://www.flowjo.com/solutions/seqgeq) - GUI based, similar to FlowJo (flow cytometry tool), paid software

# Data processing with Seurat

## Goals of processing

Remove empty GEMs, multiplet GEMs, low quality cells

Scale/normalize data

Dimensional reduction

## Isolating high-quality cells

Cellranger has built-in filtering to remove empty GEMs

-   Helps reduce size of initial data, but insufficient for rigorous analysis

Additional, manual filtering performed in Seurat on a per-capture basis

-   Knowledge of study design and goals informs stringency

## QC metrics

Mitochondrial gene expression

-   Cells with high proportion of MT gene expression are often dying, not a good representation of normal cell state

Number of unique reads

-   Low \# of reads may not capture all cell activities

-   High \# of reads may suggest multiple cells in a single GEM (multiplets)

Number of unique genes/features

-   Considered with \# of reads as a proxy for cell complexity

-   Cells with reads coming from more features need more reads to accurately capture the relative expression of all detected features

-   A cell with many reads coming from few genes may be a technical artifact that doesn't reflect a real biological state

------------------------------------------------------------------------

Knowledge of organism, tissue, and desired cell-types are key

-   Quiescent populations may have higher \# of unique features

-   Red blood cells may have majority of reads in a few features

-   Disease studies may be interested in dying cells

-   Each capture introduces technical variation

You shouldn't just apply general filtering thresholds to every study

Do you value putative detection of subtle signals, or high-confidence analyses?

## Outlier detection approach

A good starting point can be to filter outliers from your data

Median absolute deviation (MAD) outlier detection: exclude values beyond $$n$$ MADs

$$
MAD = median(|X_i - median(X)|)
$$

Increasing the allowed number of MADs reduces the stringency. A general cutoff used by developers of the scater\^1 package is 4 MADs.

<footnote>1. [Single-Cell Analysis Toolkit for Gene Expression Data in R](https://bioconductor.org/packages/release/bioc/html/scater.html)</footnote>
