---
title: "scRNAseq Markers and Cell annotation"
author: "Me!"
date: "`r Sys.Date()`"
output: html_document
---

# 0 - Project Setup

## Load packages

```{r}
library(EnhancedVolcano)
library(harmony)
library(here)
library(Seurat)
library(tidyverse)
library(fgsea)
library(msigdbr)
set.seed(43)
source(here("05_multi-sample/05-helper_functions.R"))
```

## Downloading the necessary files

Make sure that you've downloaded a zip of the github repo https://github.com/yerkes-gencore/scRNAseqWorkshop2023 and unzip it somewhere on your computer (e.g. in your `Desktop` directory).

If you want to keep working in the directory you created last class, simply copy and paste the `05_multi-sample` folder into the base level of your existing directory.

If you'd like to start fresh in a clean working directory, you can simply work from the newly unzipped folder.

*Note: If you are comfortable using git (beyond the scope of this workshop) and are already working within a local clone of the repo, you can skip the above and simply `git pull` the repo.*

## Read in processed objects

We'll start working with objects in lists, since we may have many captures we need to integrate.
Recall that each capture should be processed independently, so it should probably be it's own object.

```{r}
obj_list <- list()
obj_list$s101_stim <- readRDS(here('saved_rds/04/s101_stim_annotated.rds'))
obj_list$s101_ctrl <- readRDS(here('saved_rds/04/s101_ctrl_annotated.rds'))

obj_list$s1488_stim <- readRDS(here('saved_rds/04/s1488_stim_annotated.rds'))
obj_list$s1488_ctrl <- readRDS(here('saved_rds/04/s1488_ctrl_annotated.rds'))

obj_list$s1016_stim <- readRDS(here('saved_rds/04/s1016_stim_annotated.rds'))
obj_list$s1016_ctrl <- readRDS(here('saved_rds/04/s1016_ctrl_annotated.rds'))
```

A helpful way to run the same command on each item in a list is to use the function `lapply()`.

The first argument is the list you want to iterate over, in our case, `obj_list`. Second argument is a function you would like to apply and all remaining arguments are passed to that function. Here's an example:

```{r}
lapply(obj_list, # list to run the function on each element of
       DimPlot, # function to run on each element of list
       group.by = 'seurat_clusters') # argument to pass to the function we're running on each element
```

You can also apply a custom function using lapply. In fact, you can define a function in the lapply call itself. For example:
```{r}
# You could also put a multi-line statement within the brackets, just like any custom function
lapply(obj_list, function(x) { 
  DimPlot(x, group.by = "seurat_clusters")
})
```


Recall that UMAP coordinates will differ from capture to capture, making it difficult to interpret differences among samples and across conditions. 

```{r}
lapply(obj_list, DimPlot)
```

# 1 - Working with multiple captures

## Merging objects

We will start by merging our objects into one object. We also add a metadata
column so each cell is tagged with it's original capture. With this information,
we can always split the captures back out, and we can see the success of our 
integration. 

```{r}
## adding metadata based on the names we gave the objects when loading them
for (capture.ID in names(obj_list)){
  obj_list[[capture.ID]]$capID <- capture.ID
}
## Seurat extends the base merge() function to work on lists of Seurat objects
## Adding cell IDs prevents possible collisions in barcodes across captures
obj_merged <- merge(obj_list[[1]], y = obj_list[2:length(obj_list)], add.cell.ids = names(obj_list))
```

Our merged object should now have all the cells and genes from the split objects

```{r}
lapply(obj_list, dim)

print('obj_merged')
dim(obj_merged)
```

We want to identify shared populations of cells between captures and analyze
changes in expression for similar populations. However, simply merging captures 
is often insufficient to analyze them together. Variation between captures is 
often stronger than variation between celltypes, causing merged objects to 
cluster by capture or condition rather than shared celltypes.

```{r}
obj_merged <- NormalizeData(obj_merged, normalization.method = 'LogNormalize')
obj_merged <- ScaleData(obj_merged, verbose = F)
obj_merged <- FindVariableFeatures(obj_merged, selection.method = "vst", nfeatures = 2000, verbose = F)
obj_merged <- RunPCA(obj_merged, npcs = 30, verbose = F)
obj_merged <- RunUMAP(obj_merged, dims = 1:30, verbose = F)
obj_merged <- FindNeighbors(obj_merged, dims = 1:30, verbose = F)
obj_merged <- FindClusters(obj_merged, resolution = 0.8, verbose = F)
```

Look at the clusters in the merged object. Would we call this "well-integrated"? 

```{r, fig.width=12, fig.height=6}
DimPlot(obj_merged, group.by = 'capID') + DimPlot(obj_merged, group.by = 'consensus_labels')
```

```{r, fig.height=12, fig.width=8}
DimPlot(obj_merged, split.by = 'capID', ncol = 2) + theme(aspect.ratio = 1)
```


## Seurat integration

Choose which features to use when integrating

```{r}
anchor_features <- SelectIntegrationFeatures(obj_list)
## This is just a list of genes
head(anchor_features)
```

Find integration anchors using the RPCA method. What are the pros and cons of the RPCA method? Why do you think we've chosen it for this coding exercise?
```{r}
integration_anchors <- FindIntegrationAnchors(obj_list,
                                              anchor.features = anchor_features,
                                              scale = FALSE,
                                              normalization.method = 'LogNormalize',
                                              reduction = 'rpca',
                                              dims = 1:30)
```

Integrate data
```{r}
obj_rpca <- IntegrateData(anchorset = integration_anchors,
                                normalization.method = 'LogNormalize')
```

```{r}
obj_rpca <- obj_rpca %>%
  ScaleData() %>%
  RunPCA() %>%
  RunUMAP(dims = 1:30) %>%
  FindNeighbors(dims = 1:30) %>%
  FindClusters()
```

How does this compare to merging? Do we see more overlap of captures?

```{r, fig.height=6, fig.width=12}
DimPlot(obj_rpca, group.by = 'capID') + DimPlot(obj_rpca, group.by = 'consensus_labels')
```


```{r, fig.height=12, fig.width=8}
DimPlot(obj_rpca, split.by = 'capID', ncol = 2) + theme(aspect.ratio = 1)
```

## Harmony integration

```{r}
obj_harmony <- RunHarmony(
  obj_merged,
  group.by.vars = 'capID',
  reduction.save = 'harmony',
  max_iter = 20
)
```

```{r}
obj_harmony
obj_harmony@reductions$harmony[1:5,1:5]
dim(obj_harmony@reductions$harmony)
```

Similar to PCA, harmony reduction gives us many dimensions. We'll need to run
UMAP to make a good 2d visualization.

```{r}
obj_harmony <- RunUMAP(obj_harmony, reduction = 'harmony', dims = 1:30) %>%
  FindNeighbors(reduction = 'harmony', dims = 1:30) %>%
  FindClusters()
```

How does this compare to Seurat integration? How does it compare to merging?

```{r, fig.height=12, fig.width=8}
DimPlot(obj_harmony, group.by = 'capID') + 
  DimPlot(obj_harmony, group.by = 'consensus_labels') 
```

```{r, fig.width=12}
DimPlot(obj_harmony, split.by = 'capID', ncol = 2) + theme(aspect.ratio=1)
```


## Choose a method to move forward with

Which technique looked best to you? Select one to move forward with. 

Use ctrl-F (Windows) or command-F (Mac) to change all the following instances of `obj_placeholder` with `obj_merged`, `obj_rpca`, or `obj_harmony`.

You can remove the others from your environment if memory use is a concern.
```{r}
## Optional, remove the other objects from your environment, e.g.:
## rm(unwanted_obj)
```

## Assign integrated cluster identities

In practice, you should use the workflow from the previous session to thoroughly
evaluate markers and assign celltypes. Here we will use a truncated workflow
to quickly gloss over this and get to differential expression.

```{r}
DimPlot(obj_placeholder, label = TRUE)
```


```{r, fig.width = 9, fig.height = 10, warning = FALSE}
plotRefMapScoresFacet(obj_placeholder, label_column = "SingleR_pruned.labels", 
                      label_score_column = "SingleR_delta.next", 
                      clusters_column = "seurat_clusters")
```

```{r}
obj_placeholder$consensus_labels <- obj_placeholder$seurat_clusters
Idents(obj_placeholder) <- "consensus_labels"

obj_placeholder <- RenameIdents(
  obj_placeholder,
  "0" = "t cells",
  "1" = "monocytes",
  "2" = "t cells",
  "3" = "t cells",
  "4" = "b cells",
  "5" = "t cells",
  "6" = "NK cells",
  "7" = "monocytes",
  "8" = "Unknown",
  "9" = "b cells",
  "10" = "b cells",
  "11" = "Unknown",
  "12" = "Unknown",
  "13" = "Dendritic cells"
)
obj_placeholder$consensus_labels <- Idents(obj_placeholder)
```

```{r}
DimPlot(obj_placeholder, group.by = 'consensus_labels', label = TRUE)
```


# 2 - Differential expression

Now that we have our captures successfully integrated/merged and we've identified
shared cell populations, we can perform differential expression across conditions.

First, we'll add metadata for our conditions in case it hasn't already been
added.

```{r}
## We can just take the end of the capID 
obj_placeholder$condition <- stringr::str_sub(obj_placeholder$capID, start = -4)
```

Then run Seurat's FindMarkers, focusing on a specific cluster and the comparison
of ctrl and stim cells in that cluster

```{r}
cluster_de <- FindMarkers(obj_placeholder,
                           group.by = 'condition',
                           ident.1 = 'stim',
                           ident.2 = 'ctrl',
                           subset.ident = 'monocytes', ## Chose an Ident here
                           test.use = 'wilcox',
                           logfc.threshold = -Inf, # 
                           min.pct = 0.1,
                           only.pos = FALSE,
                           assay = 'RNA')
```

DGE is calculated where a positive log-fold change (LFC) indicates greater expression 
in ident.1, and a gene with negative LFC is more highly expressed in ident.2.

```{r}
cluster_de
```

## Violin plots

It's a good idea to visualize the observed values to see if they match what you
can visibly detect.

```{r, fig.width=10}
VlnPlot(obj_placeholder, features = head(rownames(cluster_de),3),
        group.by = 'seurat_clusters', split.by = 'condition',
        ncol = 1)
```

Seurat's violin plot function is a little limited, so we have a custom function
to focus on the cells of interest (see the helper script for more details).

```{r, fig.height=4, warning = FALSE}
gcoreVlnPlot(obj_placeholder, 
             genes = head(rownames(cluster_de),3),
             subset = '5', 
             grouping_var = 'condition', groups = c('ctrl', 'stim'),
             filter_zeros = FALSE)
```

## Volcano plot

Volcano plots are a popular visualization to show an overview of transcriptional
change between states. Lots of activity in the upper left and right quadrants 
suggest strong change.

```{r, fig.width=8, fig.height=6}
## See helper functions script for more details 
generateVolcanoPlot(cluster_dge, labels = 20, pCutoff = 1e-05, FCcutoff = log2(1.3))
```


# 3 - GSEA: Gene set enrichment analysis

Gene Set Enrichment Analysis (GSEA) is a computational method that determines 
whether an a priori defined set of genes shows statistically significant,
concordant differences between two biological states (e.g. phenotypes). 
GSEA takes a list of genes ranked by some metric (often the 
magnitude/significance, and direction of differential expression between two 
groups) and a set of pathway definitions. Enrichment can be loosely interpreted 
as collective up or down regulation of a pathway. Pathways can be any set of 
genes, including custom ones for your study. We often uses genesets from the 
[Molecular Signatures Database](https://www.gsea-msigdb.org/gsea/msigdb/index.jsp).
For GSEA performed on differential expression outputs, a positive enrichment 
score indicates the pathway is enriched in the first group
of a differential expression comparison, while a negative enrichment score means
the pathway is enriched in the second group.

For example, if differentially expressed genes are calculated between a 
'treatment' and 'control' group, where positive log-fold change
indicates the gene is more highly expressed in the 'treatment' group, a 
positive enrichment score means the pathway is enriched in the 'treatment' group. 

You can read more details on GSEA [here](https://www.pnas.org/doi/10.1073/pnas.0506580102)

```{r}
## you can see the available datasets on the website, or via the command:
msigdbr::msigdbr_collections()
## Some collections are available for other species, see the command:
msigdbr::msigdbr_species()
```

```{r}
hallmark_pathways <- msigdbr(species = "human", category = "H")
hallmark_pathways <- hallmark_pathways %>% 
  select(gene_symbol, gs_name) %>%
  unstack()
hallmark_pathways$HALLMARK_INFLAMMATORY_RESPONSE
```

```{r}
prepare_gsea_stats <- function(dge_result){
  ## Why +.Machine$double.xmin? Some p-values are below the smallest possible 
  ## value in R, which R rounds to 0, which causes some issues with calculations.
  ## We add the smallest possible value to all p-values to make them non-zero.
  stat <- dge_result$avg_log2FC * -log10(dge_result$p_val_adj + .Machine$double.xmin)
  ## There are differing opinions on the best stat to use for GSEA with single-cell
  ## data. Another approach is to use only the logFC value, which assumes some
  ## shrinkage due to the estimated significance. However, seurat's base
  ## fold-change calculation is pretty rudimentary, and this approach should
  ## help weight genes based on the consistency of differential expression. 
  stat <- setNames(stat, nm = rownames(dge_result))
  ## GSEA expects a named sorted list
  stat <- sort(stat)
  stat
}
```

```{r}
gsea_results <- fgsea::fgseaSimple(pathways = hallmark_pathways, 
                                   stats = prepare_gsea_stats(cluster_dge),
                                   nperm = 1000)
```

```{r}
arrange(gsea_results, pval)
```

